### PR TITLE
BUG: implement _switch_ds for PhasePlot

### LINKED ISSUE #4291: BUG: ParticlePhasePlot does not redraw when _switch_ds is called
### Bug report

**Bug summary**

The cookbook recipe ([Time Series Movie](https://yt-project.org/doc/cookbook/complex_plots.html#time-series-movie)) for animating plots suggests using `plot._switch_ds(new_ds)` to update the plot as part of the animation function. This does not work for `ParticlePhasePlot`s (and likely `PhasePlot`s in general).  Relevant to #4288.

**Code for reproduction**

```python
# This is copied directly from the cookbook with the modifications to create
# a semi-comparable ParticlePhasePlot animation
from matplotlib import rc_context
from matplotlib.animation import FuncAnimation

import yt

ts = yt.load("GasSloshingLowRes/sloshing_low_res_hdf5_plt_cnt_*")

#plot = yt.SlicePlot(ts[20], "z", ("gas", "density"),window_size=2)
plot = yt.ParticlePhasePlot(ts[0], ("gas","x"), ("gas", "y"), ("gas","density"), figure_size=2, x_bins=100, y_bins=100)
plot.set_unit(("gas","x"),"Mpc") # To match the SlicePlot units
plot.set_unit(("gas","y"),"Mpc")

plot.set_zlim(("gas", "density"), 8e-29, 3e-26)

fig = plot.plots[("gas", "density")].figure

# animate must accept an integer frame number. We use the frame number
# to identify which dataset in the time series we want to load
def animate(i):
    ds = ts[i]
    plot._switch_ds(ds)

animation = FuncAnimation(fig, animate, frames=len(ts))

# Override matplotlib's defaults to get a nicer looking font
with rc_context({"mathtext.fontset": "stix"}):
    animation.save("animation.mp4")
```

**Actual outcome**

https://user-images.githubusercontent.com/1833991/211651361-c05f464a-f339-4597-9880-576b085c3724.mp4
Note that the figure is updating, you can add a timestamp for example that will update correctly in each frame, but the underlying plot is not being replotted or something. 

**Expected outcome**

https://user-images.githubusercontent.com/1833991/211652283-2db846ee-76ec-4e91-a6cf-4b8c4ea3c8bd.mp4

This was generated by generating each plot individually, saving them to separate files, and then using `imageio` to combine them into an mp4:
```python
import os
import tempfile
import imageio
import yt

ts = yt.load("GasSloshingLowRes/sloshing_low_res_hdf5_plt_cnt_*")

with tempfile.TemporaryDirectory() as tempdirname:
    filelist = []
    # generate temporary figures
    for idx,ds in enumerate(ts):
        z_field = ("gas","density")
        p = yt.ParticlePhasePlot(ds, ("gas","x"), ("gas","y"), z_field,figure_size=2, x_bins=100, y_bins=100) 
        p.set_unit(("gas","x"),"Mpc")
        p.set_unit(("gas","y"),"Mpc")
        p.set_zlim(("gas", "density"), 8e-29, 3e-26)
        filename = os.path.join(tempdirname,f"ds{idx}.png")
        filelist.append(filename)
        p.save(filename)
    # build movie from files
    with imageio.get_writer('./animationFromSavedFiles.mp4', mode='I') as writer:
        for filename in filelist:
            image = imageio.imread(filename)
            writer.append_data(image)
```

**Version Information**

  * Operating System: linux
  * Python Version: 3.10.6
  * yt version: main repo branch, 4.1.x
  * Other Libraries (if applicable): NA

yt installed from source

### PR DESCRIPTION
**Pull request by neutrinoceros**
## PR Summary

fixes #4291
In short, the basic implementation of `_switch_ds` doesn't work correctly for `PhasePlot` because it doesn't regenerate the underlying profile data.

### CODE REVIEW

File: yt/visualization/profile_plotter.py [hunk 1]
```diff
@@ -982,6 +983,45 @@ def _initialize_instance(
...
+        name = old_object._type_name
+        kwargs = {n: getattr(old_object, n) for n in old_object._con_args}
+        kwargs["center"] = getattr(old_object, "center", None)
+        if data_source is not None:
+            if name != "proj":
+                raise RuntimeError(
+                    "The data_source keyword argument "
+                    "is only defined for projections."
+                )
+            kwargs["data_source"] = data_source
```
Comment (neutrinoceros) [inline yt/visualization/profile_plotter.py:998]: this can probably be dropped completely, but I'll need a hand to confirm that I'm reading this right.

**Comment by neutrinoceros** [comment]
I added a test that fails on main and pass here. I'll open for review when CI is green again.
Note that I'm deliberately leaving #4288 aside for now, and focused on making this patch backportable.

**Comment by mtryan83** [comment]
So this does seem to fix the "nothing is happening aspect", but it doesn't seem to generate the images I would expect. It's a little tricky to show, but if I compare the static image of the last dataset and the last frame of the animation, both generated with the same plot command, I see differences.
```Python
ts = yt.load("GasSloshingLowRes/sloshing_low_res_hdf5_plt_cnt_*")

def makeplot(ds):
    col_field = ("gas","density")
    plot = yt.ParticlePhasePlot(ds, ("gas","x"), ("gas", "y"), col_field, figure_size=2, x_bins=100, y_bins=100)
    plot.set_unit(("gas","x"),"Mpc")
    plot.set_unit(("gas","y"),"Mpc")
    plot.set_zlim(col_field, 8e-29, 3e-26)
    return plot, col_field # returning the color field for id purposes

def makeplotSlice(ds):
    col_field = ("gas","density")
    plot = yt.SlicePlot(ds, "z", col_field,window_size=2)
    plot.set_zlim(col_field, 8e-29, 3e-26)
    return plot, col_field

def make_animation_directly(dss,plotcommand):
    plot,col_field = plotcommand(dss[0])
    fig = plot.plots[col_field].figure
    # animate must accept an integer frame number. We use the frame number
    # to identify which dataset in the time series we want to load
    def animate(i):
        ds = dss[i]
        print(f"Drawing frame {i}", end = "\r")
        plot._switch_ds(ds)
        
    animation = FuncAnimation(fig, animate, frames=len(dss))
    if save_filename is not None:
        animation.save(save_filename)
    return animation

# Generate plots
plot, _ = makeplot(ts[-1])
plot.show()
ani = make_animation_directly(ts, makeplot)
ani
```
plot: 
image

last frame:
image

Note the differing pixel resolution for example. Differences show up earlier too, the final frame/dataset is just the easiest to reproduce.  (And it's much more obvious with the simulation data I'm working on, though I'm not sure why). I don't see a difference if I use the makeplotSlice command instead.  

This might relate to a missing `_recreate_frb` attribute. If I try 
```Python
plot,_,_ = makeplot(ts[0])
plot._switch_ds(ts[-1])
plot.show()
plot,_,_ = makeplot(ts[-1])
plot.show()
```
I get an `AttributeError: 'ParticlePhasePlot' object has no attribute '_recreate_frb'` in line 159 of `yt/yt/visualization/_commons.py` whereas it works fine (and produces identical images) if I use
```Python
plot,_,_ = makeplotSlice(ts[0])
plot._switch_ds(ts[-1])
plot.show()
plot,_,_ = makeplotSlice(ts[-1])
plot.show()
```
image

image

**Comment by neutrinoceros** [comment]
Thanks a lot for the detailed review, it's really helpful. I can see however that the complete patch is certainly going to grow larger than a reasonable backport, so I won't aim for including it in the next bugfix release just yet.

**Comment by neutrinoceros** [comment]
Okay le me dive a little bit deeper


> This might relate to a missing _recreate_frb attribute. If I try
...
I get an AttributeError: 'ParticlePhasePlot' object has no attribute '_recreate_frb' in line 159 of yt/yt/visualization/_commons.py

I think that this is symptomatic of a design weakness we've had in the code base for a while: many class methods in `PlotContainer` and its hierarchy rely on `@(in)validate_(plot|figure|data)` decorators, which supposedly avoids duplicating code, so in principle every class using them should conform to common protocols, but currently they refuse to`. As a result, these decorators end up knowing way too much about their user classes and try desperately to adapt to each of them, but this is evidence that this model doesn't work well (at least it doesn't scale).
There is evidence that this problem was noticed before, as this old comment shows:
https://github.com/yt-project/yt/blob/e43ec1da01938410374f31b3a77441c641438f5a/yt/visualization/_commons.py#L156-L157

So I think what's needed here is a clean refactor of these decorators and their user classes. I may experiment with that within this PR, but I expect this may be a rabbit hole, so I'll switch to draft for now.

**Comment by matthewturk** [comment]
In lieu of that, I would suggest we consider switching to using traitlets.

**Comment by neutrinoceros** [comment]
I can't in good faith say I want to give it a try. At least I need to be convinced that it will make my job easier rather than harder, given that I currently don't know the framework at all.

**Comment by matthewturk** [comment]
That's fair.  Let's talk in lower-latency about it, and if I can help with it, because I have a considerable amount of experience doing it, and I am keen to revisit my attempts at it from a few years ago.

**Comment by neutrinoceros** [comment]
Vitals are not good on this PR. I'll take it out of the 4.2 milestone so it doesn't block the release

**Comment by mtryan83** [comment]
I just finished my defense and can work on this again, is there anything I can do to help here or with #4301?

**Comment by neutrinoceros** [comment]
Hi @mtryan83, first of: congrats !

I still think an important refactor is needed to address this properly, and I already have too many plates spinning, so unfortunately I do not think that I can reasonably get back to it any time soon.
If anyone would like to use my work as a starting point for the refactor I've hinted at, I'll be happy to serve as a reviewer and/or provide some more guidelines. I'd also be happy to see this superseded by #4301 if we can get it going.

**Comment by neutrinoceros** [comment]
I'll close this to clear the backlog but I reaffirm that anyone wanting to push this has my support and I'd be happy to help.

**Pull request closed without merging**